packages/peregrine/lib/talons/AddressBookPage/useAddressBookPage.js

```diff
@@ -56,14 +90,93 @@ export const useAddressBookPage = (props = {}) => {
     }, [isRefetching, setPageLoading]);
 
     const handleAddAddress = useCallback(() => {
-        alert('TODO!');
+        formApi.reset();
+
+        setActiveEditAddress(null);
+        // Hide all previous errors when we open the dialog.
+        setDisplayError(false);
+        setIsDialogOpen(true);
+    }, [formApi]);
+
+    const handleEditAddress = useCallback(
+        address => {
+            const formAddress = {
+                ...address,
+                default_shipping:
+                    address.default_shipping || customerAddresses.length === 1
+            };
+            formApi.reset();
+            formApi.setValues(formAddress);
+            setActiveEditAddress(formAddress);
+            // Hide all previous errors when we open the dialog.
+            setDisplayError(false);
+            setIsDialogOpen(true);
+        },
+        [customerAddresses, formApi]
+    );
+
+    const handleCancelDialog = useCallback(() => {
+        setIsDialogOpen(false);
     }, []);
 
-    const customerAddresses =
-        (customerAddressesData &&
-            customerAddressesData.customer &&
-            customerAddressesData.customer.addresses) ||
-        [];
+    const handleConfirmDialog = useCallback(
+        async formValues => {
+            if (isDialogEditMode) {
+                try {
+                    await updateCustomerAddress({
+                        variables: {
+                            addressId: activeEditAddress.id,
+                            updated_address: formValues
+                        },
+                        refetchQueries: [{ query: getCustomerAddressesQuery }]
+                    });
+
+                    setIsDialogOpen(false);
```

Comment:
I suppose this means refetch isn't blocking, which is why we get a flash of that stale data. It also appears to behave this way in the Checkout experience, not sure if I missed that or it's just amplified by a slow backend. This is fine as-is, I don't think we should dive into optimistic UI just yet.

Reply:
> I don't think we should dive into optimistic UI just yet.

My thoughts exactly. `awaitRefetchQueries` solved my needs for now 👍

Reply:
Incredible, nice find!